### PR TITLE
Add Tekton Design Principles

### PR DESCRIPTION
**Pull request by jerop**
The goal of these principles is to provide a basis for design choices and help in making tradeoffs. The design principles are reusability, simplicity, flexibility and conformance. 

The principles were initially proposed and discussed in [this google doc](https://docs.google.com/document/d/1_iqwslxAOx-SPYosnTFS5kOZcmxcBWpuIULZHQHP85Q/edit?usp=sharing) that's visible to members of tekton-dev@. Thank you for the contributions!

Related issue: [Add contributor and reviewer expectations](https://github.com/tektoncd/community/pull/133).

### CODE REVIEW

File: design-principles.md
Comment (vdemeester) [inline design-principles.md:4]: Another way to say this (or a complement to this) would be : when implemented a feature, make sure you tried with existing features so that you can make sure you have a clear reason to add this feature.

File: design-principles.md
Comment (vdemeester) [inline design-principles.md:16]: I am not sure I fully *understand* the word `plugin` in that context.

File: design-principles.md
Comment (vdemeester) [inline design-principles.md:17]: I am not sure I completely follow. I think this needs an example or something — I understand what we are getting at but it's not that clear to me when reading

File: design-principles.md
Comment (vdemeester) [inline design-principles.md:20]: Isn't this highly specific ? (on "coupled" to `tektoncd/pipeline`)

Comment (dibyom) [reply design-principles.md:20]: Applies to triggers too 😛 

I think it might be useful to link to the design doc on whether to template or not which provides the basis for our recommendation!

File: design-principles.md
Comment (jerop) [reply design-principles.md:16]: I understood this to mean that implementations can be made in the step level or DSL level and that people can now use custom tasks -- it was from @ImJasonH so maybe he can clarify further

File: design-principles.md
Comment (jerop) [reply design-principles.md:25]: I understand kubernetes-specific here means things such as configurations for a Pod

/cc @ImJasonH

File: design-principles.md
Comment (jerop) [reply design-principles.md]: added an example that Tekton API should not be aware of git concepts, such as branches

File: design-principles.md
Comment (jerop) [reply design-principles.md]: good catch, linked the doc :)

**Comment by bobcatfish** [comment]
I'm very excited! Any thoughts I had were addressed in the original doc so:

/lgtm

Waiting for review from @vdemeester @afrittoli  (and anyone/everyone else who would like to chime in of course) before merging

**Review by vdemeester — COMMENTED** [review]
/lgtm

File: design-principles.md
Comment (afrittoli) [inline design-principles.md:17]: I agree with this, but the step vs task/pipeline distinction is not 100% clear to me.
Does this mean for example that we want a `CloneRepo` task that includes branch logic in the step, `if git then elif snv then ...`? 
I think having tasks specifically designed and tested for each SCM, selected with conditionals (when expressions?_) etc would be a valid alternative.

Comment (bobcatfish) [reply design-principles.md:17]: that makes sense to me, maybe we can reword this as something like:

- To keep Tekton flexible, specific choices should be implemented within Tasks or at the DSL level by users, not at the level of features provided by Tekton in the Task or Pipeline APIs. Such as Tekton should not be aware of git concepts, such as branches.

File: design-principles.md
Comment (afrittoli) [inline design-principles.md:26]: NIT: it would be good to have a link here to what we mean by conformance - Jason's PR with the API spec is not merged yet thought, so we could add that later

File: design-principles.md
Comment (afrittoli) [reply design-principles.md]: @jerop thanks for the update. 
Sorry to insist on this, but I feel it would be better to be specific about the Tekton API, as @bobcatfish suggested. 
Tekton should avoid being opinionated in the Task and Pipeline *API* , and Tasks (e.g. from the catalog) should be a valid place where to be specific and opinionated.

Comment (jerop) [reply design-principles.md]: thanks for making that point clearer, updated it: https://github.com/tektoncd/community/pull/171/files#diff-0899720b7e7bf227aaa00d838f2c9e76R17

File: design-principles.md
Comment (jerop) [reply design-principles.md]: agreed, added a link to the API spec tep: https://github.com/tektoncd/community/blob/master/teps/0012-api-spec.md

**Review by vdemeester — APPROVED** [review]
/cc @afrittoli

**Comment by bobcatfish** [comment]
Hey @jereop, just realized we didn't add any links to this - can we add a link from the main README?

**Pull request merged**
